src/poll.c

```diff
@@ -0,0 +1,372 @@
+#include <fcntl.h>
+#include <pthread.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/poll.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include "common.h"
+#include "poll.h"
+#include "log.h"
+
+
+// ---------------------------------------------------------------------------
+// internal data structures
+
+/*
+ * We maintain two concurrent arrays.
+ * poll_fd_array which is the data structue given to poll and
+ * poll_data_array which contains metadata we also need.
+ */
+struct pollfd *poll_fd_array = NULL;
+struct poll_data {
+	char *path;
+	uint32_t path_hash;
+	// Last update timestamp
+	struct timeval *tv;
+	int num_checker;
+} *poll_data_array;
+
+int poll_num = 0; // Size of poll_fd_array|poll_data_array
+int poll_first_deprecated = -1; // First index in poll_fd_array|poll_data_array that can be overwritten.
+
+// Every registrator holds a void pointer to one of these.
+struct poll_check {
+	int poll_array_index;
+	// Last check timestamp.
+	struct timeval tv;
+};
+
+pthread_t poll_thread;
+
+// ---------------------------------------------------------------------------
+// poll locks
+
+// Lock poll data structure (poll_fd_array, poll_data_array, poll_num and 
+// poll_first_deprecated.
+pthread_mutex_t poll_array_mtx = PTHREAD_MUTEX_INITIALIZER;
+
+static inline int poll_array_lock() {
+	if(pthread_mutex_lock(&poll_array_mtx) == 0) {
+		return 0;
+	} else {
+		error("Unable to lock poll file list");
+		return -1;
+	}
+}
+
+static inline int poll_array_unlock() {
+	if(pthread_mutex_unlock(&poll_array_mtx) == 0) {
+		return 0;
+	} else {
+		error("Unable to unlock poll file list");
+		return -1;
+	}
+}
+
+// ---------------------------------------------------------------------------
+// internal signal handler used to interrupt poll.
+void poll_handler(int signo)
+{
+	if(signo){} // Remove compiler warning
+}
+
+// ---------------------------------------------------------------------------
+// internal methods
+int poll_time_update_nolock(struct timeval *tv) {
+	int retv = gettimeofday(tv, NULL);
+	if(retv != 0) {
+		error("Could not get current time");
+	}
+	return retv;
+}
```

Comment:
Ok. I am wiling to change it. Should I log an error at the call of `poll_time_update_nolock`?